File: integrations/aws-strands/python/src/ag_ui_strands/citations.py
"""Model citations, folded onto the assistant message they annotate.

When a model answers over documents with citations enabled, it returns the
source passages its answer came from. Strands surfaces each one as a
``CitationStreamEvent`` between the text deltas of the same turn, so a citation
arrives in the middle of the message it belongs to rather than at the end of the
run.

They ride the assistant message's own ``metadata`` under
:data:`CITATIONS_METADATA_KEY`, which is what keeps a citation attached to the
thing it annotates. A separate event stream would hand the frontend two
sequences and the job of correlating them.

Two envelopes exist across the supported Strands range, because
``CitationStreamEvent`` changed shape mid-1.x. Verified by downloading each
published wheel and reading ``strands/types/_events.py``::

    1.15.0 - 1.20.0   {"callback": {"citation": ..., "delta": ...}}
    1.21.0 - latest   {"citation": ..., "delta": ...}

Both are read here. The second is also the reason nothing in this module ever
forwards a whole event: from 1.21.0 the ``delta`` key is top level, so
``ModelStreamEvent.prepare()`` merges the live ``Agent`` -- system prompt,
message history, model config -- into the same dict. Only the citation object
itself is read out of it.

The normalisation below is written against ``citations.ts`` in the TypeScript
adapter, field for field, because the two are expected to emit the same object
for the same Bedrock response. Where they cannot agree, the difference is named
in the docstring that owns it rather than left for a reader to discover.
"""

from __future__ import annotations

import json
import logging
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)

#: Metadata key carrying the citation list on an assistant message.
#:
#: Deliberately not nested under AG-UI's reserved ``ag-ui`` key. That key is
#: reserved for the protocol's own values, and metadata merging replaces a
#: key's value wholesale rather than blending it, so a second writer under
#: ``ag-ui`` would silently destroy whatever the protocol had put there.
CITATIONS_METADATA_KEY = "citations"

#: Location kinds whose Bedrock wrapper key differs from the discriminator the
#: Strands TypeScript SDK produces for the same location.
#:
#: Only one differs today. Verified against the published SDK's
#: ``_mapBedrockCitationLocation``: Bedrock wraps a search-result location in
#: ``searchResultLocation`` and the SDK emits ``type: "searchResult"``, while
#: ``documentChar``, ``documentPage``, ``documentChunk`` and ``web`` keep their
#: names. Without this rename the two adapters would disagree on that one kind.
_LOCATION_KIND_ALIASES = {"searchResultLocation": "searchResult"}


def citation_from_event(event: Any) -> Optional[Dict[str, Any]]:
    """Return the citation object carried by ``event``, or ``None``.

    A citation is only recognised alongside its ``delta`` sibling, matching how
    Strands constructs the event. An event carrying an unpaired ``citation``
    key comes from somewhere else and belongs to the RAW fallback, which exists
    for exactly the events this adapter has no branch for.
    """
    if not isinstance(event, dict):
        return None

    for container in (event, event.get("callback")):
        if not isinstance(container, dict):
            continue
        citation = container.get("citation")
        if isinstance(citation, dict) and isinstance(container.get("delta"), dict):
            return citation

    return None


def _text_entries(value: Any) -> List[Dict[str, str]]:
    """Keep the non-empty ``{"text": ...}`` entries of a citation content list.

    Dropping the empty ones is load-bearing for cross-language agreement: the
    TypeScript Strands SDK coalesces a missing text to ``""`` where Python
    leaves the entry out entirely.
    """
    if not isinstance(value, list):
        return []
    return [
        {"text": entry["text"]}
        for entry in value
        if isinstance(entry, dict) and isinstance(entry.get("text"), str) and entry["text"]
    ]


def normalize_location(location: Any) -> Optional[Dict[str, Any]]:
    """Return a citation location in discriminated form, or ``None`` if empty.

    Bedrock sends the location wrapped in the key naming its kind
    (``{"documentChar": {"documentIndex": 0, "start": 1, "end": 9}}``), while
    the TypeScript Strands SDK hands its adapter the flattened, discriminated
    form (``{"type": "documentChar", ...}``). Both shapes arrive here, because
    a custom model provider can forward either, and both are reduced to the
    flattened one.

    Four rules, each mirrored in ``citations.ts``:

    * the wrapper key becomes ``type``, renamed through
      :data:`_LOCATION_KIND_ALIASES` where the SDK uses a different name;
    * fields the provider left empty or absent are dropped, matching the SDK,
      which omits a falsy ``domain`` rather than emitting it;
    * a location that ends up naming only its kind is not a location;
    * a location not discriminated by a non-empty string ``type`` is not a
      location either. Reached when a wrapper's payload is not an object
      (``{"documentChar": "0-9"}``), which would otherwise pass through as
      provider garbage and rescue a citation that names no source.

    The unwrap is structural rather than a list of known kinds, so a variant
    added after this was written still reaches a Python client. It will NOT
    reach a TypeScript one: the SDK's own Bedrock mapper logs and returns
    nothing for a kind it does not recognise, dropping the citation with it.
    That difference is upstream and cannot be normalised away here.
    """
    if not isinstance(location, dict):
        return None

    flattened: Dict[str, Any]
    if len(location) == 1:
        kind, fields = next(iter(location.items()))
        if isinstance(fields, dict) and "type" not in fields:
            flattened = {"type": _LOCATION_KIND_ALIASES.get(kind, kind), **fields}
        else:
            flattened = dict(location)
    else:
        flattened = dict(location)

    trimmed = {
        key: value
        for key, value in flattened.items()
        if value is not None and value != ""
    }

    if not isinstance(trimmed.get("type"), str) or not trimmed["type"]:
        return None
    if set(trimmed) == {"type"}:
        return None
    return trimmed


def _json_round_trip(value: Any) -> Any:
    """A deep copy of ``value``, or ``None`` if it will not survive JSON.

    Metadata rides an event that is encoded for the SSE stream, and a value
    that will not encode fails the whole stream, costing the client its
    ``TEXT_MESSAGE_END``, snapshots and ``RUN_FINISHED``. ``allow_nan=False``
    rejects NaN and +/-Infinity, which the default would emit as bare tokens
    that are not valid JSON. The except clause is broad on purpose: a deeply
    nested value raises ``RecursionError``, which is neither ``TypeError`` nor
    ``ValueError`` and would otherwise escape the guard and kill the stream the
    guard exists to protect.
    """
    try:
        return json.loads(json.dumps(value, allow_nan=False))
    except Exception:
        return None


def _describe_keys(value: Dict[str, Any]) -> str:
    """Render a dict's keys for a log line without assuming they sort.

    A malformed citation is exactly where mixed-type keys turn up, and this
    runs on the path that exists to survive one.
    """
    return ",".join(sorted(str(key) for key in value))


def normalize_citation(citation: Any, text_offset: int) -> Optional[Dict[str, Any]]:
    """Reduce a Strands citation to the wire shape, or ``None`` to drop it.

    Empty strings, empty lists and empty locations are omitted rather than
    emitted. The TypeScript SDK coalesces a missing ``source`` or ``title`` to
    ``""`` while Python leaves the key out, so keeping the empties would make
    the two bridges disagree about a citation that is identical on the provider
    side.

    A citation that names no source is dropped outright. The generated span is
    not one of those fields, and on this side it is not available at all: see
    the note below the guard.

    Serializability is checked rather than assumed: see
    :func:`_json_round_trip`.
    """
    if not isinstance(citation, dict):
        logger.warning("Dropping a citation that is not an object: %r", type(citation))
        return None

    entry: Dict[str, Any] = {}

    for field in ("title", "source"):
        value = citation.get(field)
        if isinstance(value, str) and value:
            entry[field] = value

    raw_location = citation.get("location")
    location = normalize_location(raw_location)
    if location is not None:
        entry["location"] = location
    elif raw_location:
        # Kept, minus the location. A provider that sends an untagged shape
        # still named a source; dropping the whole citation over a field this
        # adapter cannot place would lose more than it protects.
        logger.warning(
            "Omitting a citation location that is not in tagged form "
            "(keys=%s). A location must be either Bedrock's single-key "
            "wrapper or a discriminated object with a string `type`.",
            _describe_keys(raw_location)
            if isinstance(raw_location, dict)
            else type(raw_location).__name__,
        )

    source_content = _text_entries(citation.get("sourceContent"))
    if source_content:
        entry["sourceContent"] = source_content

    if not entry:
        logger.warning(
            "Dropping a citation carrying no source fields (keys=%s)",
            _describe_keys(citation),
        )
        return None

    # No ``content`` here, deliberately. The generated span is not a field of a
    # citation in either SDK: TypeScript puts it on ``CitationsDelta`` and
    # Python on ``CitationsContentBlock``, which is the assembled block rather
    # than anything the streaming path yields. Reading it off the citation
    # would be a branch that can never fire. The TypeScript adapter takes it
    # from the delta, where its SDK does supply it, so ``content`` reaches a
    # TypeScript client and not a Python one. See the README.
    # Keys captured before the offset joins them, so the log names the fields
    # the provider sent, which is what the TypeScript sibling reports too.
    described = _describe_keys(entry)
    entry["textOffset"] = text_offset

    encodable = _json_round_trip(entry)
    if encodable is None:
        logger.warning("Dropping unserializable citation (keys=%s)", described)
    return encodable


def copy_metadata(metadata: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    """An independent copy of a citation metadata object.

    The event on the wire and the assistant message retained in the snapshot
    list need separate copies. The retained message is re-emitted in every
    later snapshot of the run, so a consumer mutating the list it received on
    ``TEXT_MESSAGE_END`` would otherwise corrupt every snapshot after it.

    Every value here has already passed :func:`normalize_citation`'s strict
    round trip, so this one cannot fail on content the module produced.
    """
    if metadata is None:
        return None
    copied = _json_round_trip(metadata)
    if copied is None:
        logger.warning("Dropping a citation list that no longer encodes")
    return copied


class CitationAccumulator:
    """Collects the citations of one assistant message.

    The accumulator also owns the running text offset. It deliberately does not
    read the adapter's ``accumulated_text``: that variable is reset only when
    message snapshots are being emitted, so an offset derived from it would
    keep counting across messages whenever snapshots are off.

    A message's citations are published as a complete list every time, because
    metadata merging replaces a key's value rather than appending to it. So
    each publish carries every citation seen so far for the open message, and a
    client always holds a whole prefix rather than a fragment.
    """

    def __init__(self) -> None:
        self._items: List[Dict[str, Any]] = []
        self._unpublished = False
        self._offset = 0

    def advance(self, delta: str) -> None:
        """Record text emitted for the open message.

        Counted in UTF-16 code units, not Python characters. The number is an
        index into the message text a client holds, and every client that will
        slice with it is a browser, where string indices are UTF-16 units. The
        TypeScript adapter gets the same number for free from
        ``String.prototype.length``; matching it here is what stops the two
        bridges disagreeing on any text containing an emoji.

        Counted per character rather than by encoding the string, because a
        lone surrogate is representable in a Python ``str`` and would make
        ``encode`` raise from the text path of every run, cited or not.
        """
        if not isinstance(delta, str):
            delta = str(delta)
        self._offset += sum(2 if ord(char) > 0xFFFF else 1 for char in delta)

    def add(self, citation: Any) -> None:
        """Record a citation against the text emitted for this message so far."""
        entry = normalize_citation(citation, self._offset)
        if entry is None:
            return
        self._items.append(entry)
        self._unpublished = True

    def __len__(self) -> int:
        return len(self._items)

    def pending(self) -> Optional[Dict[str, Any]]:
        """Metadata to attach mid-stream, or ``None`` if nothing is new.

        Returning ``None`` once published is what stops every remaining text
        delta of the message from re-sending an unchanged list.
        """
        if not self._unpublished:
            return None
        self._unpublished = False
        return self.metadata()

    def metadata(self) -> Optional[Dict[str, Any]]:
        """An independent copy of every citation collected so far, or ``None``.

        A fresh copy each call, so a consumer holding an earlier publish cannot
        see a later one mutate underneath it.
        """
        if not self._items:
            return None
        copied = _json_round_trip(self._items)
        if copied is None:
            logger.warning(
                "Dropping %d citation(s) that no longer encode", len(self._items)
            )
            return None
        return {CITATIONS_METADATA_KEY: copied}

    def take(self) -> Optional[Dict[str, Any]]:
        """Metadata for the closing message, resetting for the next one.

        Both the ``TEXT_MESSAGE_END`` event and the assistant message inside
        the following ``MESSAGES_SNAPSHOT`` need this value, and they need
        separate copies of it: see :func:`copy_metadata`.

        Callers drain this at every message boundary, including boundaries
        where no message was open, so citations cannot survive into the next
        message. See :func:`discard_orphans`.
        """
        metadata = self.metadata()
        self._items = []
        self._unpublished = False
        self._offset = 0
        return metadata


def discard_orphans(citations: CitationAccumulator, context: str) -> None:
    """Drop citations collected while no assistant message was open.

    Reached when a turn produces citations and no text. There is no message for
    them to annotate, and carrying them forward would attach one message's
    sources to the next one at a meaningless offset, so they are dropped. The
    warning exists because a silent drop here is indistinguishable from the
    model never citing anything.

    Always drains, even when there is nothing to drop, so the offset resets at
    every message boundary rather than only at the ones that had citations.
    """
    dropped = len(citations)
    citations.take()
    if dropped == 0:
        return
    logger.warning(
        "Dropping %d citation(s) that arrived with no open assistant message (%s)",
        dropped,
        context,
    )

File: integrations/aws-strands/typescript/src/citations.ts
/**
 * Model citations, folded onto the assistant message they annotate.
 *
 * When a model answers over documents with citations enabled, it returns the
 * source passages its answer came from. Strands surfaces them as
 * `citationsDelta` deltas, and Bedrock emits them between the text deltas of
 * one assistant turn, so a citation arrives in the middle of the message it
 * belongs to rather than at the end of the run. Whether the citation shares a
 * content block with the text it supports is a provider detail this adapter
 * does not depend on: the Bedrock streaming path keeps them in the same block,
 * while the OpenAI Responses adapter closes the text block first.
 *
 * They ride the assistant message's own `metadata` under
 * {@link CITATIONS_METADATA_KEY}, which is what keeps a citation attached to
 * the thing it annotates. A separate event stream would hand the frontend two
 * sequences and the job of correlating them.
 *
 * The normalisation below is written against `citations.py` in the Python
 * adapter, field for field, because the two are expected to produce equal
 * objects for the same Bedrock response. Where they cannot agree, the
 * difference is named in the docstring that owns it rather than left for a
 * reader to discover.
 */

import { DEFAULT_LOGGER, type Logger } from "./logger";

/**
 * Metadata key carrying the citation list on an assistant message.
 *
 * Deliberately not nested under AG-UI's reserved `ag-ui` key. That key is
 * reserved for the protocol's own values, and metadata merging replaces a key's
 * value wholesale rather than blending it, so a second writer under `ag-ui`
 * would silently destroy whatever the protocol had put there.
 */
export const CITATIONS_METADATA_KEY = "citations";

/**
 * Location kinds whose Bedrock wrapper key differs from the discriminator this
 * SDK produces for the same location.
 *
 * Only one differs today. Verified against the published SDK's
 * `_mapBedrockCitationLocation`: Bedrock wraps a search-result location in
 * `searchResultLocation` and the SDK emits `type: "searchResult"`, while
 * `documentChar`, `documentPage`, `documentChunk` and `web` keep their names.
 * Mirrors `_LOCATION_KIND_ALIASES` in the Python adapter.
 */
const LOCATION_KIND_ALIASES: Record<string, string> = Object.assign(
  Object.create(null) as Record<string, string>,
  { searchResultLocation: "searchResult" },
);

/**
 * Where a cited passage sits in its source.
 *
 * Mirrors the SDK's own `CitationLocation` union. Kept open-ended with a
 * catch-all member because the Python adapter forwards a kind neither SDK
 * names yet, and a consumer switching on `type` should still compile against
 * one.
 */
export type AguiCitationLocation =
  | {
      type: "documentChar" | "documentPage" | "documentChunk";
      documentIndex: number;
      start: number;
      end: number;
    }
  | {
      type: "searchResult";
      searchResultIndex: number;
      start: number;
      end: number;
    }
  | { type: "web"; url: string; domain?: string }
  | { type: string; [field: string]: unknown };

/** A citation as it reaches the client, under {@link CITATIONS_METADATA_KEY}. */
export interface AguiCitation {
  /** Title of the cited source, when the provider supplies one. */
  title?: string;
  /**
   * Source identifier, typically a URL.
   *
   * Bedrock does not send one, so this is absent on the Bedrock path in both
   * bridges. The SDK's OpenAI Responses adapter fills it with the cited URL.
   */
  source?: string;
  /** The passage in the source document that supports the answer. */
  sourceContent?: { text: string }[];
  /** Where the passage sits in the source. */
  location?: AguiCitationLocation;
  /**
   * The generated text this citation supports, where the provider reports it.
   *
   * The one field the two bridges cannot always agree on. Strands reports it on
   * the delta rather than on the citation, and only some providers populate it:
   * the SDK's OpenAI Responses adapter does, Bedrock sends an empty list. The
   * Python SDK's stream shape has no equivalent field at all, so a provider
   * that supplies a generated span reaches a TypeScript client with `content`
   * and a Python client without it.
   *
   * It does not count towards a citation naming a source. A marker carrying
   * only the text it annotates points a reader at nothing, so such a citation
   * is dropped in both adapters.
   */
  content?: { text: string }[];
  /**
   * Characters of this message's text streamed when the citation arrived,
   * counted in UTF-16 code units.
   *
   * This is the only positional information available. A citation locates a
   * span in the SOURCE document and says nothing about where in the answer it
   * belongs, so placing it in the answer relies on the provider emitting a
   * citation after the text it supports.
   *
   * Counted by {@link CitationAccumulator} rather than read off the adapter's
   * `accumulatedText`, which is reset only when message snapshots are being
   * emitted and would otherwise keep counting across messages.
   */
  textOffset: number;
}

/** The `citationsDelta` payload, as the SDK shapes it. */
interface CitationsDeltaLike {
  type?: string;
  citations?: unknown;
  content?: unknown;
}

/** Render an object's keys for a log line. */
function describeKeys(value: object): string {
  return Object.keys(value).sort().join(",");
}

/**
 * Return a citation location in discriminated form, or `undefined` if empty.
 *
 * This SDK flattens the location itself, but a custom model provider can
 * forward Bedrock's wrapped shape (`{ documentChar: { ... } }`) untouched,
 * which is also what the Python adapter receives. Both shapes arrive here and
 * both are reduced to the flattened one.
 *
 * Three rules, each mirrored in `citations.py`:
 *
 * - the wrapper key becomes `type`, renamed through
 *   {@link LOCATION_KIND_ALIASES} where this SDK uses a different name;
 * - fields the provider left empty or absent are dropped, matching the SDK,
 *   which omits a falsy `domain` rather than emitting it;
 * - a location that ends up naming only its kind is not a location.
 *
 * Anything that is not an object at all is not a location either. Python drops
 * those too, which is why a falsy scalar cannot rescue a citation here.
 */
function normalizeLocation(
  location: unknown,
): AguiCitationLocation | undefined {
  if (
    typeof location !== "object" ||
    location === null ||
    Array.isArray(location)
  ) {
    return undefined;
  }

  const keys = Object.keys(location);
  let flattened: Record<string, unknown> = {
    ...(location as Record<string, unknown>),
  };

  if (keys.length === 1) {
    const fields = (location as Record<string, unknown>)[keys[0]];
    if (
      typeof fields === "object" &&
      fields !== null &&
      !Array.isArray(fields) &&
      !("type" in fields)
    ) {
      flattened = {
        type: LOCATION_KIND_ALIASES[keys[0]] ?? keys[0],
        ...(fields as Record<string, unknown>),
      };
    }
  }

  // Null-prototype: a provider key of `__proto__` would otherwise hit the
  // prototype setter, silently dropping the field and taking its prototype
  // from provider data.
  const trimmed = Object.create(null) as Record<string, unknown>;
  for (const [key, value] of Object.entries(flattened)) {
    if (value === undefined || value === null || value === "") continue;
    trimmed[key] = value;
  }

  // Anything that did not end up discriminated is not a location. Reached when
  // a wrapper's payload is not an object at all (`{ documentChar: "0-9" }`),
  // which would otherwise pass through as provider garbage and rescue a
  // citation that names no source. It is also what makes the cast below true.
  // Anything not discriminated by a non-empty string `type` is not a
  // location. Reached when a wrapper's payload is not an object at all
  // (`{ documentChar: "0-9" }`), which would otherwise pass through as provider
  // garbage and rescue a citation that names no source. It is also what makes
  // the cast below true rather than asserted.
  const trimmedKeys = Object.keys(trimmed);
  if (typeof trimmed.type !== "string" || !trimmed.type) return undefined;
  if (trimmedKeys.length === 1) return undefined;
  // Back onto a normal prototype. Belt and braces: the JSON round trip every
  // citation passes would restore it anyway, but this object is read by
  // `normalizeCitation` before that happens.
  return { ...trimmed } as AguiCitationLocation;
}

/**
 * Keep the non-empty `{ text }` entries of a citation content list.
 *
 * Dropping the empty ones is load-bearing for cross-language agreement: this
 * SDK coalesces a missing text to `""` where the Python one leaves the entry
 * out entirely.
 */
function textEntries(value: unknown): { text: string }[] {
  if (!Array.isArray(value)) return [];
  const out: { text: string }[] = [];
  for (const entry of value) {
    const text = (entry as { text?: unknown } | null)?.text;
    if (typeof text === "string" && text) out.push({ text });
  }
  return out;
}

/**
 * Deep-copy a value, or return `undefined` if it will not survive JSON.
 *
 * Metadata rides an event that is encoded for the SSE stream, and a value that
 * will not encode fails the whole stream, costing the client its
 * `TEXT_MESSAGE_END`, snapshots and `RUN_FINISHED`. The replacer rejects
 * non-finite numbers explicitly because `JSON.stringify` would quietly turn
 * them into `null`, where the Python sibling's `allow_nan=False` raises; a
 * citation that cannot be represented is dropped in both. A bigint needs no
 * arm of its own: `JSON.stringify` throws on one already.
 */
export function jsonRoundTrip<T>(value: T): T | undefined {
  try {
    return JSON.parse(
      JSON.stringify(value, (_key, v) => {
        if (typeof v === "number" && !Number.isFinite(v)) {
          throw new TypeError("non-finite number");
        }
        return v;
      }),
    ) as T;
  } catch {
    return undefined;
  }
}

/**
 * Reduce one SDK citation to the wire shape, or `undefined` to drop it.
 *
 * Empty strings, empty lists and empty locations are omitted rather than
 * emitted, so a citation identical on the provider side produces an equal
 * object in both bridges.
 *
 * A citation that names no source is dropped outright. `content` does not count
 * towards naming one: see {@link AguiCitation.content}.
 */
function normalizeCitation(
  citation: unknown,
  content: { text: string }[],
  textOffset: number,
  log: Logger,
): AguiCitation | undefined {
  if (
    typeof citation !== "object" ||
    citation === null ||
    Array.isArray(citation)
  ) {
    const kind =
      citation === null
        ? "null"
        : Array.isArray(citation)
          ? "array"
          : typeof citation;
    log.warn(
      `[@ag-ui/aws-strands] Dropping a citation that is not an object ` +
        `(got ${kind})`,
    );
    return undefined;
  }
  const source = citation as Record<string, unknown>;

  const entry: Partial<AguiCitation> = {};
  if (typeof source.title === "string" && source.title) {
    entry.title = source.title;
  }
  if (typeof source.source === "string" && source.source) {
    entry.source = source.source;
  }

  const location = normalizeLocation(source.location);
  if (location !== undefined) {
    entry.location = location;
  } else if (source.location) {
    // Kept, minus the location. A provider that sends an untagged shape still
    // named a source; dropping the whole citation over a field this adapter
    // cannot place would lose more than it protects.
    log.warn(
      `[@ag-ui/aws-strands] Omitting a citation location that is not in ` +
        `tagged form (${
          typeof source.location === "object" && !Array.isArray(source.location)
            ? `keys=${describeKeys(source.location as object)}`
            : typeof source.location
        }). A location must be either Bedrock's single-key wrapper or a ` +
        `discriminated object with a string \`type\`.`,
    );
  }

  const sourceContent = textEntries(source.sourceContent);
  if (sourceContent.length > 0) entry.sourceContent = sourceContent;

  if (Object.keys(entry).length === 0) {
    log.warn(
      `[@ag-ui/aws-strands] Dropping a citation carrying no source fields ` +
        `(keys=${describeKeys(source)})`,
    );
    return undefined;
  }

  if (content.length > 0) entry.content = content;

  const encodable = jsonRoundTrip({ ...entry, textOffset } as AguiCitation);
  if (encodable === undefined) {
    log.warn(
      `[@ag-ui/aws-strands] Dropping unserializable citation ` +
        `(keys=${describeKeys(entry)})`,
    );
  }
  return encodable;
}

/**
 * An independent copy of a citation metadata object.
 *
 * The event on the wire and the assistant message retained in the snapshot list
 * need separate copies. The retained message is re-emitted in every later
 * snapshot of the run, so a consumer mutating the list it received on
 * `TEXT_MESSAGE_END` would otherwise corrupt every snapshot after it.
 *
 * Every value here has already passed {@link jsonRoundTrip}, so this one cannot
 * fail on content the module produced.
 */
export function copyCitationMetadata(
  metadata: Record<string, AguiCitation[]> | undefined,
  log: Logger = DEFAULT_LOGGER,
): Record<string, AguiCitation[]> | undefined {
  if (metadata === undefined) return undefined;
  const copied = jsonRoundTrip(metadata);
  if (copied === undefined) {
    log.warn(
      `[@ag-ui/aws-strands] Dropping a citation list that no longer encodes`,
    );
  }
  return copied;
}

/**
 * Collects the citations of one assistant message.
 *
 * The accumulator also owns the running text offset. It deliberately does not
 * read the adapter's `accumulatedText`: that variable is reset only when
 * message snapshots are being emitted, so an offset derived from it would keep
 * counting across messages whenever snapshots are off.
 *
 * A message's citations are published as a complete list every time, because
 * metadata merging replaces a key's value rather than appending to it. So each
 * publish carries every citation seen so far for the open message, and a client
 * always holds a whole prefix rather than a fragment.
 */
export class CitationAccumulator {
  private items: AguiCitation[] = [];
  private unpublished = false;
  private offset = 0;

  constructor(private readonly log: Logger = DEFAULT_LOGGER) {}

  /**
   * Record text emitted for the open message.
   *
   * `String.prototype.length` counts UTF-16 code units, which is what a browser
   * client will slice with. The Python adapter counts the same units rather
   * than its own characters, so the two agree on text containing an emoji.
   */
  advance(delta: string): void {
    this.offset += delta.length;
  }

  /**
   * Record the citations in a `citationsDelta`, or return false if the delta is
   * not one.
   *
   * A single delta can carry several citations. `content` is a property of the
   * delta rather than of each citation, so every citation in it gets the same
   * generated span.
   *
   * Returning true claims the delta even when every citation in it was
   * unusable, because the delta IS a citation delta and the adapter has a
   * branch for it. Each drop is warned about individually rather than handing
   * the payload to the RAW fallback, which exists for events with no branch.
   */
  add(delta: unknown): boolean {
    const candidate = delta as CitationsDeltaLike | null;
    if (candidate?.type !== "citationsDelta") return false;

    const content = textEntries(candidate.content);
    if (!Array.isArray(candidate.citations)) {
      this.log.warn(
        `[@ag-ui/aws-strands] Ignoring a citationsDelta whose citations field ` +
          `is ${candidate.citations === undefined ? "absent" : "not an array"}`,
      );
      return true;
    }
    if (candidate.citations.length === 0) {
      this.log.warn(
        `[@ag-ui/aws-strands] Ignoring a citationsDelta carrying no citations`,
      );
      return true;
    }
    for (const citation of candidate.citations) {
      const entry = normalizeCitation(citation, content, this.offset, this.log);
      if (entry === undefined) continue;
      this.items.push(entry);
      this.unpublished = true;
    }
    return true;
  }

  /**
   * Metadata to attach mid-stream, or `undefined` if nothing is new.
   *
   * Returning `undefined` once published is what stops every remaining text
   * delta of the message from re-sending an unchanged list.
   */
  pending(): Record<string, AguiCitation[]> | undefined {
    if (!this.unpublished) return undefined;
    const metadata = this.metadata();
    // Cleared only once the copy succeeded, so a build that failed is retried
    // on the next delta rather than being swallowed with the flag. Defensive
    // rather than covered: every entry already survived the identical round
    // trip on the way in, so nothing this module produces can fail here.
    if (metadata !== undefined) this.unpublished = false;
    return metadata;
  }

  /**
   * An independent copy of every citation collected so far, or `undefined`.
   *
   * A fresh copy each call, so a consumer holding an earlier publish cannot see
   * a later one mutate underneath it.
   */
  metadata(): Record<string, AguiCitation[]> | undefined {
    if (this.items.length === 0) return undefined;
    const copied = jsonRoundTrip(this.items);
    if (copied === undefined) {
      // Unreachable on content this module produced: every entry already
      // passed the same round trip on the way in. Handled rather than asserted
      // away so a future change that admits a raw value cannot ship a
      // `{ citations: undefined }` payload.
      this.log.warn(
        `[@ag-ui/aws-strands] Dropping ${this.items.length} citation(s) that ` +
          `no longer encode`,
      );
      return undefined;
    }
    return { [CITATIONS_METADATA_KEY]: copied };
  }

  /** How many citations are held for the open message. */
  get size(): number {
    return this.items.length;
  }

  /**
   * Metadata for the closing message, resetting for the next one.
   *
   * Both the `TEXT_MESSAGE_END` event and the assistant message inside the
   * following `MESSAGES_SNAPSHOT` need this value, and they need separate
   * copies of it: see {@link copyCitationMetadata}.
   *
   * Callers drain this at every message boundary, including boundaries where no
   * message was open, so citations cannot survive into the next message. See
   * {@link discardOrphanCitations}.
   */
  take(): Record<string, AguiCitation[]> | undefined {
    const metadata = this.metadata();
    this.items = [];
    this.unpublished = false;
    this.offset = 0;
    return metadata;
  }
}

/**
 * Drop citations collected while no assistant message was open.
 *
 * Reached when a turn produces citations and no text. There is no message for
 * them to annotate, and carrying them forward would attach one message's
 * sources to the next one at a meaningless offset, so they are dropped. The
 * warning exists because a silent drop here is indistinguishable from the model
 * never citing anything.
 *
 * Always drains, even when there is nothing to drop, so the offset resets at
 * every message boundary rather than only at the ones that had citations.
 */
export function discardOrphanCitations(
  citations: CitationAccumulator,
  context: string | undefined,
  log: Logger,
): void {
  const dropped = citations.size;
  citations.take();
  if (dropped === 0) return;
  log.warn(
    `[@ag-ui/aws-strands] Dropping ${dropped} citation(s) that arrived with no ` +
      `open assistant message (${context})`,
  );
}

/**
 * Spread helper so an event carries `metadata` only when there is something to
 * say. `JSON.stringify` already omits an explicit `metadata: undefined`, but
 * the object is also handed to in-process consumers and to the protobuf
 * encoder, where a present-but-undefined key is not the same as an absent one.
 */
export function citationMetadata(
  metadata: Record<string, AguiCitation[]> | undefined,
): { metadata?: Record<string, AguiCitation[]> } {
  return metadata === undefined ? {} : { metadata };
}
